repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+    val https = if (localhost.getHttpsPort > 0) List(new Port("https", localhost.getHttpsPort)) else List.empty[Port]
+    val localPorts = http ++ https
+
+    destination match {
+      case endpoint: DestinationEndpoint =>
+        Some(buildEndpoint(localhost, localPorts, endpoint, uri, servletRequest))
+      case cluster: DestinationCluster =>
+        // TODO: Remove this when routing to a cluster is removed.
+        domains.get().map(_.getDomain(cluster.getCluster.getId)) match {
+          case Some(clusterWrapper) =>
+            val routableNode = clusterWrapper.getNextNode
+            val port = if (HttpsProtocol.equalsIgnoreCase(cluster.getProtocol)) routableNode.getHttpsPort else routableNode.getHttpPort
+            Some(new URL(cluster.getProtocol, routableNode.getHostname, port, cluster.getRootPath + uri))
+          case None =>
+            logger.warn("No routable node for domain: " + cluster.getId)
+            None
+        }
+      case _ =>
+        logger.warn(s"Unknown destination type: ${destination.getClass.getName}")
+        None
+    }
+  }
+
+  def buildEndpoint(localhost: Node, localPorts: List[Port], endpoint: DestinationEndpoint, uri: String, request: HttpServletRequest): URL = {
+    val protocol = endpoint.getProtocol
+    val endpointPort = endpoint.getPort
+    val endpointHost = endpoint.getHostname
+    val path = StringUriUtilities.concatUris(endpoint.getRootPath, uri)
+
+    val portUrl = if (!StringUtils.isBlank(protocol)) {
+      val portNum = if (endpointPort <= 0) {
+        localPorts.find(_.getProtocol.equalsIgnoreCase(protocol)) match {
+          case Some(port) => port.getNumber
+          case _ => 0
+        }
+      } else {
+        endpointPort
+      }
+      new Port(protocol, portNum)
+    } else {
+      val port = new Port(request.getScheme, request.getLocalPort)
+      if (localPorts.contains(port)) {
+        port
+      } else {
+        throw new MalformedURLException("Cannot determine destination port.")
+      }
+    }
+    // IF the endpoint hostname is blank, THEN it is local
+    val hostnameUrl = if (StringUtils.isBlank(endpointHost)) localhost.getHostname else endpointHost
+
+    new URL(portUrl.getProtocol, hostnameUrl, portUrl.getNumber, path)
+  }
+
+  def doDispatch(servletRequest: HttpServletRequestWrapper,
+                 servletResponse: HttpServletResponse,
+                 locationUrl: URL,
+                 uri: String,
+                 targetContext: ServletContext,
+                 routeDest: RouteDestination,
+                 dispatcher: RequestDispatcher,
+                 destination: Destination): Unit = {
+    servletRequest.setScheme(locationUrl.getProtocol)
+    servletRequest.setServerName(locationUrl.getHost)
+    servletRequest.setServerPort(locationUrl.getPort)
+    servletRequest.setRequestURI(locationUrl.getPath)
+
+    requestHeaderService.setVia(servletRequest)
```

Comment:
Why is this a service instead of a helper of some sort? I know it's not your doing, but we should make a story to change it.

Reply:
This was made a service because it is not static and has to listen to the `system-model.cfg.xml`.

Reply:
Wow. This is another one that could be collapsed into this.
Two more classes OBE'd.